crates/sui-core/src/transaction_driver/effects_certifier.rs

```diff
@@ -65,30 +71,36 @@ impl EffectsCertifier {
                 None => (None, None),
             },
         };
+
+        // List of peers to try getting full effects from.
+        let mut retrier = TransactionRetrier::new(authority_aggregator);
+
         let (acknowledgments_result, mut full_effects_result) = join!(
-            self.wait_for_acknowledgments_with_retry(
+            self.wait_for_acknowledgments(
                 authority_aggregator,
                 tx_digest,
                 consensus_position,
                 options,
             ),
-            async move {
+            async {
+                // No need to send a full effects request if it is already provided.
                 if let Some(full_effects) = full_effects {
                     return Ok(full_effects);
                 }
-                self.get_full_effects_with_retry(
-                    authority_aggregator,
-                    tx_digest,
-                    consensus_position,
-                    options,
-                )
-                .await
+                let (name, client) = retrier.next_target()?;
+                current_target = name;
+                self.get_full_effects(client, tx_digest, consensus_position, options)
+                    .await
             },
         );
 
+        // If certification of transaction effects failed, it should be the error from this function,
+        // instead of failure to get individual full effects.
         let certified_digest = acknowledgments_result?;
 
-        // Retry until full effects digest matches the certified digest.
+        // Retry until there is a valid full effects that matches the certified digest.
+        // This loop terminates when there are enough (f+1) non-retriable errors when getting full effects,
```

Comment:
It makes sense to terminate after f+1 non-retriable errors. However unless there is some poor categorization of errors (or something has gone really wrong with the network), then it seems by the time we get to this part of the code that we won't ever have f+1 non-retriable errors as the line before would have returned an error first.